src/middleware/devtools.ts

```diff
@@ -61,107 +74,190 @@ export const devtools =
-          }
-        } else if (
-          message.type === 'DISPATCH' &&
-          message.payload?.type === 'COMMIT'
-        ) {
-          connection.init(api.getState())
-        } else if (
-          message.type === 'DISPATCH' &&
-          message.payload?.type === 'IMPORT_STATE'
-        ) {
-          const actions = message.payload.nextLiftedState?.actionsById
-          const computedStates =
-            message.payload.nextLiftedState?.computedStates || []
-
-          computedStates.forEach(
-            ({ state }: { state: PartialState<S> }, index: number) => {
-              const action = actions[index] || 'No action found'
-
-              if (index === 0) {
-                connection.init(state)
-              } else {
-                savedSetState(state)
-                connection.send(action, api.getState())
-              }
-            }
-          )
-        }
-      })
-      connection.init(initialState)
+      isRecording = true
     }
+
+    let initialState = fn(api.setState, get, api)
+    extension!.init(initialState)
+
+    extension!.subscribe((message: any) => {
+      switch (message.type) {
+        case 'ACTION':
+          if (!api.dispatchFromDevtools) return
+          if (typeof api.dispatch !== 'function') return
+          return parseJsonThen(message.payload, api.dispatch as any)
+
+        case 'DISPATCH':
+          switch (message.payload.type) {
+            case 'RESET':
+              setStateFromDevtools(initialState, true)
+              return extension!.init(initialState)
+
+            case 'COMMIT':
+              return extension!.init(api.getState())
+
+            case 'ROLLBACK':
+              return parseJsonThen<S>(message.state, (state) => {
+                setStateFromDevtools(state, true)
+                extension!.init(state)
+              })
+
+            case 'JUMP_TO_STATE':
+            case 'JUMP_TO_ACTION':
+              return parseJsonThen<S>(message.state, (state) => {
+                setStateFromDevtools(state, true)
+              })
+
+            case 'IMPORT_STATE':
+              let { nextLiftedState } = message.payload
+              let lastComputedState =
+                nextLiftedState.computedStates.at(-1)?.state
+              if (!lastComputedState) return
+              setStateFromDevtools(lastComputedState, true)
+              extension!.send(null, nextLiftedState)
+              return
+
+            case 'PAUSE_RECORDING':
+              return (isRecording = !isRecording)
+          }
+          return
+      }
+    })
+
     return initialState
   }
+
+const parseJsonThen = <T>(stringified: string, f: (parsed: T) => void) => {
+  let parsed: T | undefined
+  let didParse = true
+  try {
+    parsed = JSON.parse(stringified)
+  } catch (e) {
+    console.error(
+      '[zustand devtools middleware] Could not parse the received json',
+      e
+    )
+    didParse = false
+  }
+  if (didParse) f(parsed as T)
```

Comment:
nit: can we do this in the try-clause and remove `didParse`?

Reply:
I think of `parseJsonThen` as a `JSON.parse` replacement. With `JSON.parse` it's the user's responsibility to catch errors (which they can forget) and log the error message. With `parseJsonThen` the user doesn't have to think about all this, not to mention it makes the code less repetitive. 

`didParse` is need because `stringified` could be `"undefined"` in that case `if (parsed !== undefined) f(parsed)` would not work as `f` should have been called with `undefined`, it's true we probably won't have a case where `stringified` is `"undefined"` but still it's good to be complete.

Moreover in #662 `parseThenJson` would be needed as it's type will be `<T>(stringified: StringifiedJson<T>, f: (parsed: T) => void) => void` where `StringifiedJson<T>` would be an opaque type, which would make parsing type-safe while `JSON.parse` is not.

Let me know if you still would prefer try-catch without `didParse` :)

Reply:
if `stringified` is `undefined`, JSON.parse will throw.
But, I realizes that my proposition can hide thrown error from `f`, so let's keep as is.

Reply:
> if stringified is undefined, JSON.parse will throw.

Ah yes I mistakenly thought `JSON.stringify(undefined)` is `"undefined"` (just like it is for null, strings, numbers, etc). In that case I'll remove `didParse`.